enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,357 @@
+So only when the node plugin finishes the registration process with NRI and device plugin, the scheduler
+admits the application pod.
+
+Once a container under the pod requests and gets access to shared cpus, new environment variable named
+`OPENSHIFT_SHARED_CPUS=<CPU-IDs>` will be present under the container's environment variables.
+
+The mixed-cpu-node-plugin itself is essentially a pod that runs as a DaemonSet on the worker nodes.
+cluster-node-tuning-operator (NTO) will be modified to support the deployment of the DaemonSet and additional
+resources (ServiceAccount, RBAC resources, SecurityContextConstraint, etc.)
+NTO will also be responsible to watch, monitor and report the node-plugin's DaemonSet status.
+Additionally, since the plugin is part of OCP infrastructure pods, NTO annotates the pods with:
+```yaml
+target.workload.openshift.io/management: {
+  "effect":"PreferredDuringScheduling"
+}
+```
+See https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md for details.
+
+Since both the components are related,
+mixed-cpu-node-plugin [source code](https://github.com/openshift-kni/mixed-cpu-node-plugin) would be vendored under NTO.
+By having different main.go + container entrypoint, NTO image can be also used as the
+mixed-cpu-node-plugin image. 
+This way saves the overhead of having a separate container image with all the maintenance cost that comes with it.
+
+In order to activate NRI in container-runtime, NTO creates MachineConfig to enable NRI in 
+the container-runtime configuration:
+```yaml
+[crio.nri]
+enable_nri = true
+```
+
+The plugin is completely optional and is off by default.
+
+No changes are needed in the container-runtime, because NRI hooks already supported by the runtime
+and mixed-cpu-node-plugin implementing the NRI interface.
+This minimizes risks and helps avoid regression on
+a container-runtime component (CRI-O in OpenShift case).
+
+_"A picture is worth a thousand words"_
+![mixed-cpu-node-plugin](mixed-cpu-node-plugin.jpg)
+
+### API
+
+Extend PerformanceProfile with new `workloadHints` named `mixedCpus` in order to enable
+the feature.
+
+```yaml
+workloadHints:
+  mixedCpus: true
+```
+
+Specify the shared cpus via annotation:
+
+```yaml
+annotations:
+    mixed-cpu.openshift.io/shared-cpus: "2-3"
+```
+
+The CPU set value should follow the 
+[cpuset](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpuset) conventions.
+It defines a set of CPUs that can be allocated to guaranteed pods/containers that still require
+non-exclusive ones.
+
+The CPU set must not overlap with `spec.cpu.reserved` or `spec.cpu.isolated` and NTO
+should return an error in case it does.
+
+The reason for using an annotation and not new API field 
+is because there is a coming feature named shared-partition
+that uses similar convention i.e., shared cpus, and would be 
+enabled via NTO as well.
```

Comment:
Adding an annotation based API means you need to support it forever, and it is unstructured and you also can't validate the input, you should create an API field.

How will this proposal and the future feature shared-partition interact with one another? Are they achieving the same goal?

Reply:
Hi @JoelSpeed thank you for your comments, the reason for adding an annotation, and I explained that in the doc, is because we want to use a future API filed that would be populated by shared-partition #1421 feature. 
shared-partition exposes a `cpu.shared` API, and mixed-cpu uses that for setting up the shared cpus.

Given your feedback, I suggested decoupling mixed-cpu and shared-partition by having a dedicated API for enablement/disablement each of the features, regardless of the `cpu.shared` filed. 
mixed-cpu already suggests `workloadHint` filed  for this purpose exactly.